### DISCUSSION TITLE
ProgressPlugin has much times 1 for the percentage

### DISCUSSION THREAD

**Discussion by Idletalk** [webpack/webpack#12996] [Q&A]
Here i use the `webpack.ProgressPlugin({handler(percentage, message, ...args){}})` to show the progress of the webpack build. but when i output the percentage and message, i find there is much times in percentage 1 in build like this 

    build [=================== ] 99% cache begin idle
    build [=================== ] 99% cache begin idle
    build [====================] 100%  
    build [=================== ] 99% cache shutdown
    build [=================== ] 99% cache shutdown
    build [====================] 100%

and then i use the server here is more output like this
    build [=================== ] 99% cache begin idle
    build [=================== ] 99% cache begin idle
    build [====================] 100% 
    build [=================== ] 99% cache shutdown                                                                                                                
    build [=================== ] 99% cache shutdown
    build [====================] 100%  
    build [=================== ] 99% end closing watch compilation
    build [=================== ] 99% end closing watch compilation

this may cause confusion. Is it Reasonable?

**Comment by alexander-akait** [comment]
What is dev server options and how you run dev server?

**Reply by Idletalk** [reply]
here is the webpack.config.js
// webpack.config.js
const path = require("path");
const HtmlWebpackPlugin = require("html-webpack-plugin");
const webpack = require("webpack");

 module.exports = {
  entry: path.join(__dirname, "/src/index.js"), // 入口文件
  output: {
    path: path.join(__dirname, "/dist"), //打包后的文件存放的地方
    filename: "bundle.js", //打包后输出文件的文件名
  },
  module: {
    rules: [
      {
        test: /\.css$/,
        use: ["style-loader", "css-loader"],
      },
    ],
  },
  plugins: [
    new webpack.ProgressPlugin({
      handler(percentage, message, ...args) {
        // console.log('webpack building...')
        // readline.clearLine(process.stdout, 1)
        const step = "=".repeat(Math.floor(percentage * 20));
        const space = " ".repeat(20 - Math.floor(percentage * 20));
        const indentation = " ".repeat(4);
        console.log(
          `${indentation}build ` +
            `[${step}${space}]` +
            ` ${Math.floor(percentage * 100)}% ${message} ${args}`
        );
        // readline.moveCursor(process.stdout, 0, -1)
      },
    }),
    new HtmlWebpackPlugin({
      template: "./src/index.html",
    }),
  ],
  mode: "development",
  devServer: {
    port: "8088", // 设置端口号为8088
    hot: true,
  },
};



ant here is the output 
   build [=================== ] 99% cache store build dependencies
    build [=================== ] 99% cache store build dependencies
    build [=================== ] 99% cache begin idle
    build [=================== ] 99% cache begin idle
    build [====================] 100%  
    build [=================== ] 99% cache shutdown
    build [=================== ] 99% cache shutdown
    build [====================] 100%  
here is the package.json
   "html-webpack-plugin": "^5.0.0",
    "style-loader": "^2.0.0",
    "webpack": "^5.0.0",
    "webpack-cli": "^4.0.0",
    "webpack-dev-server": "4.0.0-beta.0"

**Reply by alexander-akait** [reply]
Can you try `"webpack-dev-server": "4.0.0-beta.1"` and please provide command how you run dev server

**Reply by Idletalk** [reply]
> here is the webpack.config.js
> // webpack.config.js
> const path = require("path");
> const HtmlWebpackPlugin = require("html-webpack-plugin");
> const webpack = require("webpack");
> 
> module.exports = {
> entry: path.join(__dirname, "/src/index.js"), // 入口文件
> output: {
> path: path.join(__dirname, "/dist"), //打包后的文件存放的地方
> filename: "bundle.js", //打包后输出文件的文件名
> },
> module: {
> rules: [
> {
> test: /.css$/,
> use: ["style-loader", "css-loader"],
> },
> ],
> },
> plugins: [
> new webpack.ProgressPlugin({
> handler(percentage, message, ...args) {
> // console.log('webpack building...')
> // readline.clearLine(process.stdout, 1)
> const step = "=".repeat(Math.floor(percentage * 20));
> const space = " ".repeat(20 - Math.floor(percentage * 20));
> const indentation = " ".repeat(4);
> console.log(
> `${indentation}build ` +
> `[${step}${space}]` +
> ` ${Math.floor(percentage * 100)}% ${message} ${args}`
> );
> // readline.moveCursor(process.stdout, 0, -1)
> },
> }),
> new HtmlWebpackPlugin({
> template: "./src/index.html",
> }),
> ],
> mode: "development",
> devServer: {
> port: "8088", // 设置端口号为8088
> hot: true,
> },
> };
> 
> ant here is the output
> build [=================== ] 99% cache store build dependencies
> build [=================== ] 99% cache store build dependencies
> build [=================== ] 99% cache begin idle
> build [=================== ] 99% cache begin idle
> build [====================] 100%
> build [=================== ] 99% cache shutdown
> build [=================== ] 99% cache shutdown
> build [====================] 100%
> here is the package.json
> "html-webpack-plugin": "^5.0.0",
> "style-loader": "^2.0.0",
> "webpack": "^5.0.0",
> "webpack-cli": "^4.0.0",
> "webpack-dev-server": "4.0.0-beta.0"

above is the commond:'npm run build' ant 'build : webpack'.

**Reply by Idletalk** [reply]
> Can you try `"webpack-dev-server": "4.0.0-beta.1"` and please provide command how you run dev server

both the same output between 4.0.0-beta.1 and 4.0.0-beta.0

**Reply by Idletalk** [reply]
> > Can you try `"webpack-dev-server": "4.0.0-beta.1"` and please provide command how you run dev server
> 
> both the same output between 4.0.0-beta.1 and 4.0.0-beta.0

when i run commond: 'npm run dev' and "dev: webpack-dev-server". 
there will output 
build [=================== ] 99% cache begin idle
build [=================== ] 99% cache begin idle
build [====================] 100%
build [=================== ] 99% cache shutdown
build [=================== ] 99% cache shutdown
build [====================] 100%

**then i use the command+z to stop the server, they will output the message below** 

   build [=================== ] 99% cache shutdown
    build [=================== ] 99% cache shutdown
    build [====================] 100%  
    build [=================== ] 99% end closing watch compilation
    build [=================== ] 99% end closing watch compilation

**Reply by alexander-akait** [reply]
Update webpack-cli to the latest version

**Reply by Idletalk** [reply]
> webpack-cli

  "name": "webpack-cli",
  "version": "4.5.0",
  "description": "CLI for webpack & friends",

**Reply by Idletalk** [reply]
> > webpack-cli
> 
> "name": "webpack-cli",
> "version": "4.5.0",
> "description": "CLI for webpack & friends",

  "name": "webpack",
  "version": "5.28.0",
  "author": "Tobias Koppers @sokra",

**Reply by alexander-akait** [reply]
Please create reproducible test repo

**Reply by Idletalk** [reply]
> Please create reproducible test repo

https://github.com/Idletalk/demo1
 thanks

**Reply by alexander-akait** [reply]
Yep, it is expected, we intercept hooks, so multiple plugins can run the same hooks and report the same percentage

**Reply by Idletalk** [reply]
but when the webpack build complete，and percentage change 1，then the cache will result the percentage to 0.99 and change to 1 again. is it expected？

**Reply by alexander-akait** [reply]
Yep, some plugins can add more actions so we change percentage from 1 to 0.99 in this case
